book/internals/wrap_gtk_widget.nim

```diff
@@ -0,0 +1,201 @@
+If the constructor requires parameters, add fields with their values to your Widget and access the values from the implicitly available `widget` variable and its `val/has<Fieldname>` fields. 
+
+For examples, see the [beforeBuild docs](https://can-lehmann.github.io/owlkettle/book/internals/hooks/before_build_hook.html) or search for `beforeBuild:` in [widgets.nim](https://github.com/can-lehmann/owlkettle/blob/main/owlkettle/widgets.nim).
+
+### *See your example in action!*
+Once you're at this step you should be able to compile your example and see your widget in action!
+Run it with `nim r --path:. ./examples/path/to/your/example.nim` from the base dir of your Owlkettle repository clone.
+
+Use your example to manually test your widget as you add features to it in the later section.
+
+Remember to add signal-handler procs with bodies to the example, once you have added signal event listeners.
+That way you can see your application react more to various inputs.
+
+For a short example, look at Owlkettle's [counter-example](https://github.com/can-lehmann/owlkettle/blob/main/examples/counter.nim).
+
+
+### *Add fields to the Widget*
+Beyond just creating the widget and providing event-listeners, GTK may provide options to further customize a widget.
+
+E.g. you may be able to show or hide values, change their positioning, enable or disable them etc. by updating certain fields at runtime after the widget was constructed.
+
+Take a look at the functions of your widget (and the functions of its parent classes) in the GTK docs, specifically anything that isn't a getter.
+If there's procs to add or set a property of the widget, try to add that feature to your wrapped widget.
+If the proc requires a parameter of type `GtkWidget` or an array/list of `GtkWidget`, add a field of type `Widget`/`seq[Widget]` to your wrapped widget instead of `GtkWidget`. 
+
+At minimum, try to wrap all features exposed by your widget directly and possibly its direct parent (e.g. for `Gtk.Scale` everything from there as well as its parent `Gtk.Range`)
+
+Follow and repeat the following steps to add a feature:
+- Wrap the GTK functions as explained in earlier sections
+- Add a field to your widget for every parameter required by the wrapped GTK functions.
+  If a field already exists because you use it during initialization with the constructor, then you don't need to add a new field.
+- Add a `property` hook for every field on your widget that is *not* of type `Widget` and that you have a wrapped GTK function for.
+  The hook should do nothing but call the wrapped GTK function with values from the implicitly available `state` variable.
+- Add a `(build, update)` hook for every field on your widget that *is* of type `Widget`.
+  The hook should do nothing but call `state.updateChild` or `state.updateChildren` with the wrapped GTK functions for adding/removing a `GtkWidget`.
```

Comment:
```suggestion
```

This is only true in a very limited number of cases. Often you need to implement custom logic, though that is way to complex for this section.